solarextrapolation/classes.py

```diff
@@ -0,0 +1,333 @@
+    where magnetogram observations are generally taken.
+    """
+    def __init__(self, map_data, **kwargs):
+        """Method for creating a preprocessor object, using a sunpy map.
+        """
+        # Add some type checking, we want a map object, check for .unit attribute.
+        self.map_input = map_data
+        self.routine = kwargs.get('preprocessor_routine', type(self))
+        self.meta = { 'original_map_header': self.map_input.meta, 'preprocessor_routine': self.routine }
+        self.filepath = kwargs.get('filepath', None)
+
+    def _preprocessor(self):
+        """
+        Method running the and returning a sunpy map.
+        For tracability this should add entries into the metadata that
+        include any parameters used for the given run.
+        """
+        map_output = sunpy.map.Map(self.map_input.data, self.meta)
+        return map_output
+
+    def preprocess(self):
+        
+        """
+        Method to be called to run the preprocessor.
+        Times the process and saves output where applicable.
+        """
+        tim_start = time.time()
+        map_output = self._preprocessor()
+        tim_duration = time.time() - tim_start
+
+        map_output.meta['preprocessor_start_time'] = tim_start
+        map_output.meta['preprocessor_duration'] = tim_duration
+
+        if self.filepath:
+            map_output.save(self.filepath)
+        return map_output
+    
+
+
+class Extrapolators(object):
+    """
+    Common class for all 3D vector field extrapolation routines.
+    Each routine, created by building a subclass, will have wildly varying
+    capabilities and input arguments so this have been left intentionally
+    minimal.
+    The primary method to override is extrapolation(), the primary method to
+    call is extrapolate() which will both call extrapolation() and save the
+    result if a filepath argument is given.
+    """
+
+    def __init__(self, map_magnetogram, **kwargs):
+        """Construct an extrapolator using the given 2D map.
+        """
+        self.map_boundary_data = map_magnetogram
+        self.meta = { 'boundary_1_meta': self.map_boundary_data.meta }
+        self.z = kwargs.get('z', 10)
+        self.filepath = kwargs.get('filepath', None)
+        self.routine = kwargs.get('extrapolator_routine', type(self))
+        self.meta = { 'original_map_header': self.map_boundary_data.meta, 'extrapolator_routine': self.routine }
+
+    def _extrapolation(self):
+        """
+        The routine for running an extrapolation routine.
+        This is the primary method to be edited in subclasses for specific
+        extrapolation routine implementations.
+        """
+        # Add some type checking, we want a map object, check for .unit attribute.
+        # Extrapolation code goes here.
+        arr_4d = np.zeros([0, 0, 0, 0])
+        map_output = Map3D( arr_4d, self.meta )
+        
+        return map_output
+
+    def extrapolate(self):
```

Comment:
I think the `filepath` kwarg and with it the option to turn file saving on or off would be better as an argument to this function rather than the constructor, unless you have use for it else where?